command/debug.go

```diff
@@ -0,0 +1,956 @@
+		c.logger.Info("capturing configuration state")
+
+		resp, err := client.Logical().Read("sys/config/state/sanitized")
+		if err != nil {
+			captErr := newCaptureError("config", err)
+			c.debugIndex.Errors = append(c.debugIndex.Errors, captErr)
+
+			c.logger.Error("config: error capturing config state", "error", err)
+		}
+
+		if resp != nil && resp.Data != nil {
+			collection := []map[string]interface{}{
+				{
+					"timestamp": time.Now().UTC(),
+					"config":    resp.Data,
+				},
+			}
+			if err := c.persistCollection(collection, "config.json"); err != nil {
+				c.UI.Error(fmt.Sprintf("Error writing data to %s: %v", "config.json", err))
+			}
+		}
+	}
+
+	return nil
+}
+
+// capturePollingTargets captures all dynamic targets over the specified
+// duration and interval.
+func (c *DebugCommand) capturePollingTargets(client *api.Client) error {
+	startTime := time.Now()
+	durationCh := time.After(c.flagDuration + debugDurationGrace)
+
+	// Track current interval count to show progress.
+	var idxCount, mIdxCount int
+
+	// errCh is used to capture any polling errors and recorded in the index
+	errCh := make(chan *captureError)
+
+	var serverStatusCollection []map[string]interface{}
+	var replicationStatusCollection []map[string]interface{}
+	var metricsCollection []map[string]interface{}
+	var hostInfoCollection []map[string]interface{}
+
+	// Start a goroutine to collect data as soon as it's received
+	collectCh := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case metricsEntry := <-c.metricsCh:
+				metricsCollection = append(metricsCollection, metricsEntry)
+			case serverStatusEntry := <-c.serverStatusCh:
+				serverStatusCollection = append(serverStatusCollection, serverStatusEntry)
+			case replicationStatusEntry := <-c.replicationStatusCh:
+				replicationStatusCollection = append(replicationStatusCollection, replicationStatusEntry)
+			case hostInfoEntry := <-c.hostInfoCh:
+				hostInfoCollection = append(hostInfoCollection, hostInfoEntry)
+			case <-c.doneCh:
+				close(collectCh)
+				return
+			}
+		}
+	}()
+
+	// Start capture by capturing the first interval before we hit the first
+	// ticker.
+	c.pollingWg.Add(1)
+	go c.intervalCapture(client, idxCount, startTime, errCh)
+
+	c.pollingWg.Add(1)
+	go c.metricsIntervalCapture(client, mIdxCount, errCh)
+
+	intervalTicker := time.Tick(c.flagInterval)
+	metricsIntervalTicker := time.Tick(c.flagMetricsInterval)
+	interrupted := false
+
+	// Capture at each interval, until end of duration or interrupt.
+POLLING:
+	for {
+		select {
+		case err := <-errCh:
+			c.debugIndex.Errors = append(c.debugIndex.Errors, err)
+		case <-intervalTicker:
+			idxCount++
+			c.pollingWg.Add(1)
+			go c.intervalCapture(client, idxCount, startTime, errCh)
+		case <-metricsIntervalTicker:
+			mIdxCount++
+			c.pollingWg.Add(1)
+			go c.metricsIntervalCapture(client, mIdxCount, errCh)
+		case <-durationCh:
+			break POLLING
+		case <-c.ShutdownCh:
```

Comment:
I think we are still likely to deadlock if durationCh fires while there are still inflight captures that need to send an error down. On option would be to move the error channel up to the other for loop that is running in a go routine.